tree-sitter-tips.md

```diff
@@ -0,0 +1,79 @@
+
+### :bulb: Rely on Language Specs, but not too much.
+There are language specifications (like [this one for Java](https://docs.oracle.com/javase/specs/jls/se9/html/index.html)) that comprehensively describe a language. Specs provide the lexical and syntactic structure of a language in addition to its components (ex., types, variables, declarations, etc.) and how they interconnect. This can serve as a great baseline for writing your grammar. However, mirroring the spec 100% is not always the best approach, especially when its structure does not align with the parse tree you'd like to generate using tree-sitter. In this situation, it's better to deviate from the spec and prioritize generating a good parse tree. One way to do this may be to find an LR parser for the language you're working on ([such as this one for Java](https://github.com/javaparser/javaparser)) to see how they've modeled their tree outputs. Studying tree outputs from other LR parsers can be a good way to guide your own grammar development. 
+
+### :bulb: What does a "good" parse tree look like?
+
+Here are some attributes of a "pretty" tree:
+
+:white_check_mark: Not overly verbose.
+
+:white_check_mark: Everything is outlined clearly.
+
+:white_check_mark: Communicates good information needed further along in the pipeline.
+
+![image](https://user-images.githubusercontent.com/875834/34972577-41bc12ac-fa37-11e7-9a12-a05d18b6f6e4.png)
+
+_Example of a good tree (WIP -- better examples welcome!):_
+
+```
+===
+module
+===
+
+module com.foo { }
+open module com.foo { }
+
+---
+
+(program
+  (module_declaration (module_identifier (identifier) (identifier)))
+  (module_declaration (module_identifier (identifier) (identifier))))
+```
+
+### :bulb: Improving a parse tree by making it more compact
+Here are some things that might help:
+
+**1. Aliasing** - by aliasing different things, you can maintain compact tree structure while marking differences in your grammar. _(Need better description)_
+
+**2. Making things hidden** - by putting an underscore before a rule (`_rule` vs `rule`), it will allow you to omit displaying it in the tree.
+
+### :bulb: Testing in the wild to help prioritize what to support next.
+
+Once you have developed a significant portion of the grammar, find a file from an active open source repository that follows the typical structure of the language and see how much of it can be successfully parsed by your system. Is it 40%? 60%? This can be a good way to benchmark how practical your grammar is in terms of popular usage of your language. It can also be an effective way to identify what to support and prioritize what's next.
+
+### :bulb: Knowing when to stop development
+Most languages have a long-tail of features that don't get utilized frequently. It is reasonable to prioritize developing 80% of a language and only supporting commonly used elements.
```

Comment:
I think we probably want to support 100% of a language—or, curiously, more, since we often want to support multiple versions and such—but not necessarily all in one go, if that makes sense. I.e. spread it out over multiple PRs, and/or accept that you’ll catch errors “in the wild” sometimes. Is that what you mean here?

Reply:
No, I guess I was trying to posit that while there's always more to do (given the evolving nature of languages), that for our purposes development can be paused. But you're right that I should instead say we can theoretically keep going. 

Would it be fair to say that it's worth knowing what you'll be using the output for, in order to constrain and place some guard rails around what you prioritize supporting?

Reply:
Ah—that’s an excellent point! Calling out language evolution explicitly seems like a good idea 👍

> Would it be fair to say that it's worth knowing what you'll be using the output for, in order to constrain and place some guard rails around what you prioritize supporting?

Totally 👍